### PR TITLE
Cache Docker layers with `actions/cache`

### PR DESCRIPTION
**Pull request by messense**
Let's see how it performs.

### CODE REVIEW

**Comment by messense** [comment]
Some overall numbers:

1. clean build without cache: 15m 0s
2. build with cache: 4m 7s

I think it's pretty good.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -162,13 +162,28 @@ jobs:
...
+            maturin-test-docker-${{ runner.os }}-buildx-
       - name: Build
         uses: docker/build-push-action@v2
         with:
           push: false
           tags: maturin
           load: true
-          cache-from: type=registry,ref=konstin2/maturin:buildcache
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
```
Comment (konstin) [inline .github/workflows/test.yml:179]: Does this still need `mode=max` to still work after code changes?

**Review by konstin — APPROVED** [review]
Nice, this is even faster than the docker cache

**Comment by konstin** [comment]
If we do this we should also change it in the docker publish job

File: .github/workflows/test.yml (hunk 1 above)
Comment (messense) [reply .github/workflows/test.yml:179]: I am not sure, though I've never encountered any issue in my past experience with this approach.

Comment (messense) [reply .github/workflows/test.yml:179]: I'll push a minor code change to test it.

Comment (konstin) [reply .github/workflows/test.yml:179]: It now rebuilds all dependencies, with `mode=max` it should only rebuild maturin itself

**Comment by messense** [comment]
Looks like now it's worse than registry cache. So closing.

**Comment by konstin** [comment]
This might also only be compiling maturin vs. a completly cached be, I'll trigger a new build just to check

**Comment by konstin** [comment]
Fully cached the new solution is a minute faster (https://github.com/PyO3/maturin/runs/3642983166 vs. https://github.com/PyO3/maturin/runs/3644100792), so I'd still merge this.

**Comment by konstin** [comment]
Merged in 3aee10d71aeeee80f288634173835a4a7452bb9b

**Pull request closed without merging**